Cuda accelerator

StickOnAStick opened the pull request:
Adds CUDA to accelerate matrix calculations. 


|  Cuda Enabled  | Warmup Tokens / S | Simulation Tokens/S |
| -------------- | ----------------- | ------------------- |
|  ❌            |  ~480/s          | ~350/s              |
|  ✅            |  ~2500/s         | ~1850/s              |


**Screenshots:** 
### Non-Cuda:
![image](https://github.com/user-attachments/assets/ecef9e23-c128-421e-a22d-3084fc7147d4)
### Cuda Enabled:
![image](https://github.com/user-attachments/assets/7c6528a4-922c-400a-a668-b61d74f73230)

## Test Used

Warmup Instruction Count: **20,000**
Simulation Instruction Count: **50,000**
Test Used: **649.fotnik3d_s-1B**

## Notes:

Currently the `addCuda()` method actually _**slows down**_ the transformer. 

Observed an average of **~4s** of increase in test time and **66%** decrease in Tokens/Second across 4x runs.

Spaceswag committed:
Edited json editor to print more useful
information, this is used in the JSON merge
feature for the parallel_tester

Johngibbard committed:
Added CUDA as cuda.o, use CudaFixedVectorMath::

Johngibbard committed:
Added Additional performance by using cuda Streams
for cudacpymemasync copy back to host

Johngibbard commented on Makefile:
```diff
@@ -1,19 +1,25 @@
+CXXFLAGS += --std=c++17 -Ofast -Wall -Wextra -Wshadow -Wpedantic -fexpensive-optimizations
+
+# CUDA flags
+CUDA_FLAGS = -arch=all-major -O3  -ltoir -gen-opt-lto --use_fast_math --cudadevrt static --prec-div=false --extra-device-vectorization --default-stream per-thread 
```
Added some flags to add optimizations, some of them are redundant, but I don't remember which ones

Johngibbard commented on Makefile:
```diff
@@ -33,6 +39,7 @@ clean:
 	@-$(RM) inc/cache_modules.h
 	@-$(RM) inc/ooo_cpu_modules.h
 	@-$(RM) src/core_inst.cc
+	@-$(RM) cuda.o
```
For some reason the makefile needs to clean before it will acknowledge changes to cuda.o, Im not sure why it does this but compilation & linking works correctly if you use the command "make clean" before "make -j"

Johngibbard commented on tests/weights_old.json:
Sorry, should not have pushed the weights, don't know how that happened

StickOnAStick committed:
Removed Wrapper.hh with unnecessary CudaFixedVectorMath namespace.

Updated all FixedVectorMath method definitions to be **Inline**.
This was done to prevent linker errors when NVCC creates a seperate
instance of FixedVectorMath, see [LearnCPP Inline](https://www.learncpp.com/cpp-tutorial/inline-functions-and-variables/)
for more information.

Added "USE_CUDA" definition to transformer.cc as a
quick-n-dirty way to distinguish processing strategies.

Updated transformer.cc to utilize cuda when enabled.
